@stuartwdouglas Hi Stuart, thanks for providing a quick fix, looks like the resteasy tests have been affected:
```
Caused by: java.lang.NullPointerException
	at io.quarkus.resteasy.test.security.UserResource.get(UserResource.java:19)
```

--- a/integration-tests/oidc-code-flow/pom.xml
+++ b/integration-tests/oidc-code-flow/pom.xml
@@ -32,6 +32,11 @@
             <artifactId>quarkus-micrometer</artifactId>
         </dependency>
 
+        <dependency>
+            <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-test-security</artifactId>
+            <scope>test</scope>
+        </dependency>
         <!-- test dependencies -->
         <dependency>
             <groupId>io.quarkus</groupId>
@@ -209,7 +214,9 @@
                                 </goals>
                                 <configuration>
                                     <systemPropertyVariables>
-                                        <native.image.path>${project.build.directory}/${project.build.finalName}-runner</native.image.path>
+                                        <native.image.path>
+                                            ${project.build.directory}/${project.build.finalName}-runner
+                                        </native.image.path>
                                     </systemPropertyVariables>
                                 </configuration>
                             </execution>

--- a/integration-tests/oidc-code-flow/src/main/java/io/quarkus/it/keycloak/ProtectedResource.java
+++ b/integration-tests/oidc-code-flow/src/main/java/io/quarkus/it/keycloak/ProtectedResource.java
@@ -5,6 +5,8 @@ import javax.ws.rs.GET;
 import javax.ws.rs.InternalServerErrorException;
 import javax.ws.rs.Path;
 import javax.ws.rs.QueryParam;
+import javax.ws.rs.core.Context;
+import javax.ws.rs.core.SecurityContext;
 
 import org.eclipse.microprofile.jwt.JsonWebToken;
 
@@ -39,6 +41,15 @@ public class ProtectedResource {
 
     @Inject
     RefreshToken refreshToken;
+
+    @Context
+    SecurityContext securityContext;
+
+    @GET
+    @Path("sec")
+    public String hello() {
+        return securityContext.getUserPrincipal().getName();
+    }
 
     @GET
     public String getName() {

--- a/integration-tests/oidc-code-flow/src/test/java/io/quarkus/it/keycloak/TestSecurityLazyAuthTest.java
+++ b/integration-tests/oidc-code-flow/src/test/java/io/quarkus/it/keycloak/TestSecurityLazyAuthTest.java
@@ -1,0 +1,23 @@
+package io.quarkus.it.keycloak;
+
+import static org.hamcrest.Matchers.is;
+
+import org.junit.jupiter.api.Test;
+
+import io.quarkus.test.common.http.TestHTTPEndpoint;
+import io.quarkus.test.junit.QuarkusTest;
+import io.quarkus.test.security.TestSecurity;
+import io.restassured.RestAssured;
+
+@QuarkusTest
+@TestHTTPEndpoint(ProtectedResource.class)
+public class TestSecurityLazyAuthTest {
+
+    @Test
+    @TestSecurity(user = "user1", roles = "viewer")
+    public void testWithDummyUser() {
+        RestAssured.when().get("sec").then()
+                .body(is("user1"));
+    }
+
+}

--- a/test-framework/security/src/main/java/io/quarkus/test/security/TestIdentityAssociation.java
+++ b/test-framework/security/src/main/java/io/quarkus/test/security/TestIdentityAssociation.java
@@ -26,7 +26,7 @@ public class TestIdentityAssociation extends SecurityIdentityAssociation {
         }
     }
 
-    SecurityIdentity testIdentity;
+    volatile SecurityIdentity testIdentity;
 
     /**
      * A request scoped delegate that allows the system to function as normal when
@@ -64,8 +64,15 @@ public class TestIdentityAssociation extends SecurityIdentityAssociation {
 
     @Override
     public SecurityIdentity getIdentity() {
-        if (testIdentity != null) {
-            return testIdentity;
+        //we check the underlying identity first
+        //in most cases this will have been set by the TestHttpAuthenticationMechanism
+        //this means that all the usual auth process will run, including augmentors and
+        //the identity ends up in the routing context
+        SecurityIdentity underlying = delegate.getIdentity();
+        if (underlying.isAnonymous()) {
+            if (testIdentity != null) {
+                return testIdentity;
+            }
         }
         return delegate.getIdentity();
     }

